library/rendering/item_rendering/src/main/java/org/quiltmc/qsl/rendering/item/api/client/VanillaCountLabelRenderer.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.rendering.item.api.client;
+
+import com.mojang.blaze3d.vertex.Tessellator;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.client.font.TextRenderer;
+import net.minecraft.client.render.LightmapTextureManager;
+import net.minecraft.client.render.VertexConsumerProvider;
+import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.item.ItemStack;
+
+/**
+ * A {@link CountLabelRenderer} implementation that replicates vanilla behavior.
+ */
+@Environment(EnvType.CLIENT)
+public class VanillaCountLabelRenderer implements CountLabelRenderer {
+	@Override
+	public void renderCountLabel(MatrixStack matrices, TextRenderer renderer, float zOffset, ItemStack stack, @Nullable String override) {
+		String label = getCountLabel(stack, override);
+		if (label == null) {
+			return;
+		}
+
+		matrices.push();
+		matrices.translate(0, 0, zOffset + 200);
+		// TODO figure out if we can NOT render immediately here
```

Comment:
I suppose this TODO means you haven't found a way to avoid Immediate right? That's a bit concerning.

Reply:
I guess vanilla uses Immediate here as well right? I'll have to go look, but I think that's the case.

Reply:
Yeah, this is copied from vanilla.